File: 트리/2263/2263.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void preorder(vector<int>::iterator it1, vector<int>::iterator it2, vector<int>::iterator it3, vector<int>::iterator it4)
{
   if (it2 - it1 > 1)
   {
      int n = *(it4 - 1);
      cout << n << " ";
      n = find(it1, it2, n) - it1;
      preorder(it1, it1 + n, it3, it3 + n);
      preorder(it1 + n + 1, it2, it3 + n, it4 - 1);
   }
   else if (it3 != it4)
      cout << *(it4 - 1) << " ";
}

int main()
{
   int n;
   cin >> n;
   vector<int> v1(n), v2(n);
   for (int i = 0; i < n; ++i)
      cin >> v1[i];
   for (int i = 0; i < n; ++i)
      cin >> v2[i];
   preorder(v1.begin(), v1.end(), v2.begin(), v2.end());

   return 0;
}

File: 트리/2263/2263.py
def preorder(arr1, arr2):
   if len(arr2) > 1:
      print(arr2[-1], end=' ')
      n = arr1.find(arr2[-1])
      preorder(arr1[:n], arr2[:n])
      preorder(arr1[n+1:], arr2[n:-1])
   elif len(arr2) == 1:
      print(arr2[-1], end=' ')

n = int(input())
arr1 = input().replace(' ', '')
arr2 = input().replace(' ', '')
preorder(arr1, arr2)